advanced_source/semi_structured_sparse.py

```diff
@@ -0,0 +1,618 @@
+
+Requirements
+------------
+
+-  PyTorch >= 2.1.
+-  A NVIDIA GPU with semi-structured sparsity support (Compute
+   Capability 8.0+).
+
+This tutorial is designed for beginners to semi-structured sparsity /
+sparsity in general. For users with existing 2:4 sparse models,
+accelerating ``nn.Linear`` layers for inference with
+``to_sparse_semi_structured`` is as easy as:
+
+"""
+
+import torch
+from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor
+from torch.utils.benchmark import Timer
+SparseSemiStructuredTensor._FORCE_CUTLASS = True
+
+# mask Linear weight to be 2:4 sparse
+mask = torch.Tensor([0, 0, 1, 1]).tile((3072, 2560)).cuda().bool()
+linear = torch.nn.Linear(10240, 3072).half().cuda().eval()
+linear.weight = torch.nn.Parameter(mask * linear.weight)
+
+x = torch.rand(3072, 10240).half().cuda()
+
+with torch.inference_mode():
+    dense_output = linear(x)
+    dense_t = Timer(stmt="linear(x)",
+                    globals={"linear": linear,
+                             "x": x}).blocked_autorange().median * 1e3
+
+    # accelerate via SparseSemiStructuredTensor
+    linear.weight = torch.nn.Parameter(to_sparse_semi_structured(linear.weight))
+
+    sparse_output = linear(x)
+    sparse_t = Timer(stmt="linear(x)",
+                    globals={"linear": linear,
+                             "x": x}).blocked_autorange().median * 1e3
+
+    # sparse and dense matmul are numerically equivalent
+    # On an A100 80GB, we see: `Dense: 0.870ms Sparse: 0.630ms | Speedup: 1.382x`
+    assert torch.allclose(sparse_output, dense_output, atol=1e-3)
+    print(f"Dense: {dense_t:.3f}ms Sparse: {sparse_t:.3f}ms | Speedup: {(dense_t / sparse_t):.3f}x")
+
+
+######################################################################
+# What problem does semi-structured sparsity solve?
+# -------------------------------------------------
+# 
+# The general motivation behind sparsity is simple: if there are zeros in
+# your network, you can avoid storing / doing compute with those
+# parameters. However, the specifics of sparsity are tricky. Zeroing out
+# parameters doesn’t affect the latency / memory overhead of our model out
+# of the box.
+# 
+# This is because the dense tensor still contains the pruned (zero)
+# elements, which the dense matrix multiplication kernel will still
+# operate on this elements. In order to realize performance gains, we need
+# to swap out dense kernels for sparse kernels, which skip calculation
+# involving pruned elements.
+# 
+# To do this, these kernels work on sparse matrices, which do not store
+# the pruned elements and store the specified elements in a compressed
+# format.
+# 
+# For semi-structured sparsity, we store exactly half of the original
+# parameters along with some compressed metadata about how the elements
+# were arranged.
+# 
+# .. image::
+# https://developer-blogs.nvidia.com/wp-content/uploads/2023/06/2-4-structured-sparsity-pattern.png
+# :align: center :width: 80%
+# 
+# ::
```

Comment:
This should be just a caption for the image as far as I understand. Caption syntax can be added after a blank line with 4 indents. In other words, removing line 104 and 105 should work.
```suggestion
```